src/VisualStudio/Core/Def/Implementation/InheritanceMargin/InheritanceGlyphFactory.cs

```diff
@@ -44,21 +51,36 @@ internal sealed class InheritanceGlyphFactory : IGlyphFactory
 
         public UIElement? GenerateGlyph(IWpfTextViewLine line, IGlyphTag tag)
         {
-            if (tag is not InheritanceMarginTag inheritanceMarginTag)
+            var document = _textView.TextBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
                 return null;
+            }
 
-            var membersOnLine = inheritanceMarginTag.MembersOnLine;
-            Contract.ThrowIfTrue(membersOnLine.IsEmpty);
+            var optionService = document.Project.Solution.Workspace.Services.GetRequiredService<IOptionService>();
+            // The life cycle of the glyphs in Indicator Margin is controlled by the editor,
+            // so in order to get the glyphs removed when FeatureOnOffOptions.InheritanceMarginCombinedWithIndicatorMargin is off,
+            // we need
+            // 1. Generate tags when this option changes.
+            // 2. Always return null here to force the editor to remove the glyphs.
+            var combineWithIndicatorMargin = optionService.GetOption(FeatureOnOffOptions.InheritanceMarginCombinedWithIndicatorMargin, document.Project.Language);
+            if (combineWithIndicatorMargin && tag is InheritanceMarginTag inheritanceMarginTag)
```

Comment:
can you do the tag check first and bail out immediately if it isn't?  then do the extra work in the case that it is.